From c4e13d15e8dd1df1bd827117d1a74c187ed2b3c2 Mon Sep 17 00:00:00 2001
From: Mridul Agarwal <mridulagarwal@google.com>
Date: Tue, 12 Mar 2024 10:48:16 +0000
Subject: [PATCH] Add flag to prevent non-system apps from showing overlay on
 HC UI.

Test: manual
Bug: 328561789
(cherry picked from commit 85c7c82115c8935a0b22067c76c525ffea2459e3)
(cherry picked from https://googleplex-android-review.googlesource.com/q/commit:271dcafb550877589f83ec0690bd1a5d2d83d773)
Merged-In: I7b667f51c4a0952be20cafa852c13857d57cadc2
Change-Id: I7b667f51c4a0952be20cafa852c13857d57cadc2
---
 apk/AndroidManifest.xml                                      | 2 ++
 apk/src/com/android/healthconnect/controller/MainActivity.kt | 4 ++++
 .../healthconnect/controller/data/DataManagementActivity.kt  | 5 +++++
 .../healthconnect/controller/migration/MigrationActivity.kt  | 3 +++
 .../controller/onboarding/OnboardingActivity.kt              | 3 +++
 .../controller/permissions/request/PermissionsActivity.kt    | 4 ++++
 .../controller/permissions/shared/SettingsActivity.kt        | 3 +++
 .../healthconnect/controller/route/RouteRequestActivity.kt   | 4 ++++
 8 files changed, 28 insertions(+)

diff --git a/apk/AndroidManifest.xml b/apk/AndroidManifest.xml
index 11094c77..41ccd3c1 100644
--- a/apk/AndroidManifest.xml
+++ b/apk/AndroidManifest.xml
@@ -26,6 +26,8 @@
     <uses-permission android:name="android.permission.READ_DEVICE_CONFIG" />
     <uses-permission android:name="android.permission.LAUNCH_MULTI_PANE_SETTINGS_DEEP_LINK" />
     <uses-permission android:name="android.permission.SEND_SAFETY_CENTER_UPDATE" />
+    <!-- Used to disable non system apps from showing overlays -->
+    <uses-permission android:name="android.permission.HIDE_OVERLAY_WINDOWS" />
 
     <application
         android:name=".HealthConnectApplication"
diff --git a/apk/src/com/android/healthconnect/controller/MainActivity.kt b/apk/src/com/android/healthconnect/controller/MainActivity.kt
index 5074e056..7a16e05d 100644
--- a/apk/src/com/android/healthconnect/controller/MainActivity.kt
+++ b/apk/src/com/android/healthconnect/controller/MainActivity.kt
@@ -16,6 +16,7 @@
 package com.android.healthconnect.controller
 
 import android.os.Bundle
+import android.view.WindowManager
 import androidx.activity.viewModels
 import androidx.navigation.findNavController
 import com.android.healthconnect.controller.migration.MigrationActivity.Companion.maybeRedirectToMigrationActivity
@@ -37,6 +38,9 @@ class MainActivity : Hilt_MainActivity() {
 
     public override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
+        // This flag ensures a non system app cannot show an overlay on Health Connect. b/313425281
+        window.addSystemFlags(WindowManager.LayoutParams.SYSTEM_FLAG_HIDE_NON_SYSTEM_OVERLAY_WINDOWS)
+
         setContentView(R.layout.activity_main)
 
         setTitle(R.string.app_label)
diff --git a/apk/src/com/android/healthconnect/controller/data/DataManagementActivity.kt b/apk/src/com/android/healthconnect/controller/data/DataManagementActivity.kt
index 2e8c6368..8cd613cd 100644
--- a/apk/src/com/android/healthconnect/controller/data/DataManagementActivity.kt
+++ b/apk/src/com/android/healthconnect/controller/data/DataManagementActivity.kt
@@ -19,6 +19,7 @@
 package com.android.healthconnect.controller.data
 
 import android.os.Bundle
+import android.view.WindowManager
 import androidx.activity.viewModels
 import androidx.navigation.findNavController
 import com.android.healthconnect.controller.R
@@ -41,6 +42,10 @@ class DataManagementActivity : Hilt_DataManagementActivity() {
 
     override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
+
+        // This flag ensures a non system app cannot show an overlay on Health Connect. b/313425281
+        window.addSystemFlags(WindowManager.LayoutParams.SYSTEM_FLAG_HIDE_NON_SYSTEM_OVERLAY_WINDOWS)
+
         setContentView(R.layout.activity_data_management)
 
         if (maybeRedirectIntoTwoPaneSettings(this)) {
diff --git a/apk/src/com/android/healthconnect/controller/migration/MigrationActivity.kt b/apk/src/com/android/healthconnect/controller/migration/MigrationActivity.kt
index 0cd6d05f..55015132 100644
--- a/apk/src/com/android/healthconnect/controller/migration/MigrationActivity.kt
+++ b/apk/src/com/android/healthconnect/controller/migration/MigrationActivity.kt
@@ -20,6 +20,7 @@ import android.content.Context
 import android.content.DialogInterface
 import android.content.Intent
 import android.os.Bundle
+import android.view.WindowManager
 import androidx.fragment.app.FragmentActivity
 import androidx.navigation.findNavController
 import com.android.healthconnect.controller.R
@@ -169,6 +170,8 @@ class MigrationActivity : Hilt_MigrationActivity() {
 
     override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
+        // This flag ensures a non system app cannot show an overlay on Health Connect. b/313425281
+        window.addSystemFlags(WindowManager.LayoutParams.SYSTEM_FLAG_HIDE_NON_SYSTEM_OVERLAY_WINDOWS)
         setContentView(R.layout.activity_migration)
     }
 
diff --git a/apk/src/com/android/healthconnect/controller/onboarding/OnboardingActivity.kt b/apk/src/com/android/healthconnect/controller/onboarding/OnboardingActivity.kt
index 5bda4d16..b728e303 100644
--- a/apk/src/com/android/healthconnect/controller/onboarding/OnboardingActivity.kt
+++ b/apk/src/com/android/healthconnect/controller/onboarding/OnboardingActivity.kt
@@ -4,6 +4,7 @@ import android.app.Activity
 import android.content.Context
 import android.content.Intent
 import android.os.Bundle
+import android.view.WindowManager
 import android.widget.Button
 import androidx.fragment.app.FragmentActivity
 import com.android.healthconnect.controller.R
@@ -52,6 +53,8 @@ class OnboardingActivity : Hilt_OnboardingActivity() {
 
     override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
+        // This flag ensures a non system app cannot show an overlay on Health Connect. b/313425281
+        window.addSystemFlags(WindowManager.LayoutParams.SYSTEM_FLAG_HIDE_NON_SYSTEM_OVERLAY_WINDOWS)
         setContentView(R.layout.onboarding_screen)
 
         logger.setPageId(PageName.ONBOARDING_PAGE)
diff --git a/apk/src/com/android/healthconnect/controller/permissions/request/PermissionsActivity.kt b/apk/src/com/android/healthconnect/controller/permissions/request/PermissionsActivity.kt
index 97e3f9b6..1059e95c 100644
--- a/apk/src/com/android/healthconnect/controller/permissions/request/PermissionsActivity.kt
+++ b/apk/src/com/android/healthconnect/controller/permissions/request/PermissionsActivity.kt
@@ -27,6 +27,7 @@ import android.content.pm.PackageManager.EXTRA_REQUEST_PERMISSIONS_RESULTS
 import android.os.Bundle
 import android.util.Log
 import android.view.View
+import android.view.WindowManager.LayoutParams.SYSTEM_FLAG_HIDE_NON_SYSTEM_OVERLAY_WINDOWS
 import androidx.activity.viewModels
 import androidx.fragment.app.FragmentActivity
 import com.android.healthconnect.controller.R
@@ -64,6 +65,9 @@ class PermissionsActivity : Hilt_PermissionsActivity() {
 
         setContentView(R.layout.activity_permissions)
 
+        // This flag ensures a non system app cannot show an overlay on Health Connect. b/313425281
+        window.addSystemFlags(SYSTEM_FLAG_HIDE_NON_SYSTEM_OVERLAY_WINDOWS)
+
         if (!intent.hasExtra(EXTRA_PACKAGE_NAME)) {
             Log.e(TAG, "Invalid Intent Extras, finishing")
             finish()
diff --git a/apk/src/com/android/healthconnect/controller/permissions/shared/SettingsActivity.kt b/apk/src/com/android/healthconnect/controller/permissions/shared/SettingsActivity.kt
index 8a6c9e9a..8bdcd738 100644
--- a/apk/src/com/android/healthconnect/controller/permissions/shared/SettingsActivity.kt
+++ b/apk/src/com/android/healthconnect/controller/permissions/shared/SettingsActivity.kt
@@ -38,6 +38,7 @@ import android.os.Bundle
 import android.os.Handler
 import android.os.Looper
 import android.util.Log
+import android.view.WindowManager
 import androidx.core.os.bundleOf
 import androidx.navigation.Navigation
 import androidx.navigation.findNavController
@@ -60,6 +61,8 @@ class SettingsActivity : Hilt_SettingsActivity() {
 
     public override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
+        // This flag ensures a non system app cannot show an overlay on Health Connect. b/313425281
+        window.addSystemFlags(WindowManager.LayoutParams.SYSTEM_FLAG_HIDE_NON_SYSTEM_OVERLAY_WINDOWS)
         setContentView(R.layout.activity_settings)
         setTitle(R.string.permgrouplab_health)
     }
diff --git a/apk/src/com/android/healthconnect/controller/route/RouteRequestActivity.kt b/apk/src/com/android/healthconnect/controller/route/RouteRequestActivity.kt
index 90fa35e3..b8fc4909 100644
--- a/apk/src/com/android/healthconnect/controller/route/RouteRequestActivity.kt
+++ b/apk/src/com/android/healthconnect/controller/route/RouteRequestActivity.kt
@@ -21,6 +21,7 @@ import android.health.connect.HealthConnectManager.EXTRA_EXERCISE_ROUTE
 import android.health.connect.HealthConnectManager.EXTRA_SESSION_ID
 import android.os.Bundle
 import android.util.Log
+import android.view.WindowManager
 import android.widget.Button
 import android.widget.LinearLayout
 import android.widget.TextView
@@ -70,6 +71,9 @@ class RouteRequestActivity : Hilt_RouteRequestActivity() {
     public override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
 
+        // This flag ensures a non system app cannot show an overlay on Health Connect. b/313425281
+        window.addSystemFlags(WindowManager.LayoutParams.SYSTEM_FLAG_HIDE_NON_SYSTEM_OVERLAY_WINDOWS)
+
         if (!featureUtils.isExerciseRouteEnabled()) {
             Log.e(TAG, "Exercise routes not available, finishing.")
             setResult(Activity.RESULT_CANCELED, Intent())
-- 
2.44.0.396.g6e790dbe36-goog

